### PR TITLE
[supervisor] first shot at notification API (#3038)

### PR DESCRIPTION
**Pull request by JanKoehnlein**
First shot at the notification API.

Please review. I'd be glad about any advice how to test this.

### CODE REVIEW

**Comment by akosyakov** [comment]
> Please review. I'd be glad about any advice how to test this.

I found writing unit tests is very helpful, especially because of my little knowledge of go at the beginning. You can check out for instance how we test [tasks](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/pkg/supervisor/tasks_test.go) and [ports](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/pkg/ports/ports_test.go).

You can run tests directly from an IDE via code lenses. To detect concurrency issues it is a good idea to use go test with `-- race` and `--count` options.

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:50]:
```suggestion
		return nil, status.Error(codes.ResourceExhausted, "Max number of pending notifications exceeded")
```

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:63]: because `api.NotificationService_SubscribeServer` is an interface type, hence `subscribers` should be `[]api.NotificationService_SubscribeServer` instead of `[]*api.NotificationService_SubscribeServer`

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:52]:
```suggestion
	srv.mutex.Lock()
	defer srv.mutex.Unlock()
```

This way you don't have to make sure that you Unlock the mutex on every return path

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:81]: Not sure this is what you want. You'd want to have a blocking read, i.e.
```suggestion
        select {
            case resp<-channel:
                return resp, nil
            case <-ctx.Done():
                return nil, ctx.Err()
        }
```

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:95]: ideally this would be a gRPC `status.Error`

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:109]:
```suggestion
	if pending == nil {
		log..WithField("requestId", req.RequestId).Info("Invalid or late response to notification")
		// alternatively, probably better: return an error
		return &api.RespondResult{}, nil
	}
```

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:133]:
```suggestion
	if len(indices)== 0 {
	    return
	}
```

c.f. https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

**Review by csweichel — COMMENTED** [review]
That's awesome progress 🚀

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:52]: I don't want to keep the mutex lock when blocking for a result (sounds like a deadlock situation, as the responder tries to get the same lock as well). I guess it's best to factor out the synchronized section into a separate function and use the `defer` approach in that.

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:50]: Done

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:63]: Thanks.

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:52]: Correct me if I am wrong, but if we unlock the mutex on return only, we have a blocking wait for the response while holding the lock. The responder will not be able to acquire the mutex thus we have a deadlock.

But using `defer` looks much safer, so I'll extract the locked part of this func into a separate one that uses defer to unlock.

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:81]: Correct me if I am wrong: It was blocking in my version, too, but yours terminates also when the context is closed.

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:95]: Addressed

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:109]: Done

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:133]: Done

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [reply components/supervisor/pkg/supervisor/notification.go]: You are correct of course. The main difference is that in the select version above we also terminate on context closure, hence reduce the chance to leak go routines

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [reply components/supervisor/pkg/supervisor/notification.go:55]: Probably a regression, but the error code is `Internal` again

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:60]: You could move this decision to `notifySubscribers` which would always return a channel. If there are no actions, it would simply write `&api.NotifyResponse{}` to the channel and close it immediately.

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:82]: leftover comment

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:119]:
```suggestion
	pending, ok := srv.pendingNotifications[req.RequestId]
	if ok {
```

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:131]:
```suggestion
	delete(srv.pendingNotifications, req.RequestId)
	pending.responseChannel <- req.Response
	close(pending.responseChannel)
```

Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:131]: This write might block if the context of the original "Notify" was canceled. See comment above.

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:89]: This channel should have a capacity of 1. Otherwise, if the context of the original `Notify` call canceled the `Respond` call will block on the channel write. 1 is enough because said `Respond` call will close and remove the channel from the map afterwards.

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:148]:
```suggestion
	log.WithField("count", len(indicies)).Warn("unsubscribing stale subscribers")
```

File: components/supervisor/pkg/supervisor/notification.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification.go:159]: better: https://github.com/golang/go/wiki/SliceTricks#filter-in-place

File: components/supervisor/pkg/supervisor/notification_test.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/notification_test.go:1]: Didn't look at the tests in detail yet. Two things that come in handy for those kinds of tests are the `mock` package we usually generate for gRPC API, and writing table driven tests (e.g. [tasks_test.go](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/pkg/supervisor/tasks_test.go))

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go]: OK

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go:131]: OK

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go]: Yup

File: components/supervisor/pkg/supervisor/notification.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification.go]: Did the change, but it doesn't look much easier due to the fact that I have to find the index in the `indices` slice.

File: components/supervisor/pkg/supervisor/notification_test.go
Comment (JanKoehnlein) [reply components/supervisor/pkg/supervisor/notification_test.go:1]: These tests are more about the order of notifications/subscriptions that about a matrix of test data. The individual tests are pretty different in structure as we have to handle different blocking scenarios. I think it's easier to leave them as they are.

**Comment by JanKoehnlein** [comment]
Switching this back to draft, as it's going to need more adaptions when we implement the VS Code side

**Review by akosyakov — APPROVED** [review]
It worked nicely for me!

**Review by csweichel — APPROVED** [review]
LGTM

**Pull request merged**
